File: go/pkg/antfly/src/sim/router.go
// Copyright 2026 Antfly, Inc.
//
// Licensed under the Elastic License 2.0 (ELv2); you may not use this file
// except in compliance with the Elastic License 2.0. You may obtain a copy of
// the Elastic License 2.0 at
//
//     https://www.antfly.io/licensing/ELv2-license
//
// Unless required by applicable law or agreed to in writing, software distributed
// under the Elastic License 2.0 is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// Elastic License 2.0 for the specific language governing permissions and
// limitations.

package sim

import (
	"fmt"
	"net/http"
	"net/http/httptest"
	"net/url"
	"sync"
)

type routeEntry struct {
	handler   http.Handler
	available bool
}

// Router is an in-memory HTTP transport for simulator scenarios.
type Router struct {
	mu     sync.RWMutex
	routes map[string]routeEntry
}

func NewRouter() *Router {
	return &Router{
		routes: make(map[string]routeEntry),
	}
}

func (r *Router) Register(rawURL string, handler http.Handler) error {
	key, err := routeKey(rawURL)
	if err != nil {
		return err
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	r.routes[key] = routeEntry{handler: handler, available: true}
	return nil
}

func (r *Router) Unregister(rawURL string) error {
	key, err := routeKey(rawURL)
	if err != nil {
		return err
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	delete(r.routes, key)
	return nil
}

func (r *Router) SetAvailable(rawURL string, available bool) error {
	key, err := routeKey(rawURL)
	if err != nil {
		return err
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	entry, ok := r.routes[key]
	if !ok {
		return fmt.Errorf("route %s not found", rawURL)
	}
	entry.available = available
	r.routes[key] = entry
	return nil
}

func (r *Router) RoundTrip(req *http.Request) (*http.Response, error) {
	key := req.URL.Scheme + "://" + req.URL.Host

	r.mu.RLock()
	entry, ok := r.routes[key]
	r.mu.RUnlock()
	if !ok || !entry.available || entry.handler == nil {
		return nil, fmt.Errorf("route %s unavailable", key)
	}

	clone := req.Clone(req.Context())
	clone.RequestURI = req.URL.RequestURI()
	clone.Host = req.URL.Host

	recorder := httptest.NewRecorder()
	entry.handler.ServeHTTP(recorder, clone)
	resp := recorder.Result()
	resp.Request = req
	return resp, nil
}

func routeKey(rawURL string) (string, error) {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return "", fmt.Errorf("parse route URL %q: %w", rawURL, err)
	}
	if parsed.Scheme == "" || parsed.Host == "" {
		return "", fmt.Errorf("route URL %q must include scheme and host", rawURL)
	}
	return parsed.Scheme + "://" + parsed.Host, nil
}

File: zig/pkg/antfly/src/raft/state_machine/router.zig
// Copyright 2026 Antfly, Inc.
//
// Licensed under the Elastic License 2.0 (ELv2); you may not use this file
// except in compliance with the Elastic License 2.0. You may obtain a copy of
// the Elastic License 2.0 at
//
//     https://www.antfly.io/licensing/ELv2-license
//
// Unless required by applicable law or agreed to in writing, software distributed
// under the Elastic License 2.0 is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// Elastic License 2.0 for the specific language governing permissions and
// limitations.

const std = @import("std");
const raft_engine = @import("raft_engine");
const read_state_observer_mod = @import("read_state_observer.zig");

pub const RoutedStateMachine = struct {
    metadata_group_id: ?raft_engine.core.types.GroupId = null,
    metadata_state_machine: raft_engine.runtime.storage_iface.StateMachine,
    data_state_machine: raft_engine.runtime.storage_iface.StateMachine,
    read_state_observer: ?read_state_observer_mod.ReadStateObserver = null,

    pub fn stateMachine(self: *RoutedStateMachine) raft_engine.runtime.storage_iface.StateMachine {
        return .{
            .ptr = self,
            .vtable = &.{
                .prepare_snapshot = prepareSnapshot,
                .build_snapshot = buildSnapshot,
                .apply_ready = applyReady,
            },
        };
    }

    fn prepareSnapshot(
        ptr: *anyopaque,
        group_id: raft_engine.core.types.GroupId,
        applied_index: raft_engine.core.types.Index,
    ) !?raft_engine.runtime.storage_iface.SnapshotSource {
        const self: *RoutedStateMachine = @ptrCast(@alignCast(ptr));
        const target = if (self.metadata_group_id != null and group_id == self.metadata_group_id.?)
            self.metadata_state_machine
        else
            self.data_state_machine;
        return try target.prepareSnapshot(group_id, applied_index);
    }

    fn buildSnapshot(ptr: *anyopaque, alloc: std.mem.Allocator, group_id: raft_engine.core.types.GroupId) !?[]u8 {
        const self: *RoutedStateMachine = @ptrCast(@alignCast(ptr));
        const target = if (self.metadata_group_id != null and group_id == self.metadata_group_id.?)
            self.metadata_state_machine
        else
            self.data_state_machine;
        return try target.buildSnapshot(alloc, group_id);
    }

    fn applyReady(
        ptr: *anyopaque,
        group_id: raft_engine.core.types.GroupId,
        snapshot: ?raft_engine.core.types.Snapshot,
        committed_entries: []const raft_engine.core.Entry,
        read_states: []const raft_engine.core.ReadState,
    ) !void {
        const self: *RoutedStateMachine = @ptrCast(@alignCast(ptr));
        const target = if (self.metadata_group_id != null and group_id == self.metadata_group_id.?)
            self.metadata_state_machine
        else
            self.data_state_machine;
        try target.applyReady(group_id, snapshot, committed_entries, read_states);
        if (self.read_state_observer) |observer| {
            if (read_states.len > 0) {
                try observer.onReadStates(group_id, read_states);
            }
        }
    }
};
